### ISSUE TITLE
Requirements conflict: Requirement.parse('torch==1.7.1')

### ISSUE THREAD

**Issue by bjornfor** [DavHau/mach-nix#238]
`torch==1.7.1` is newer than the mach-nix-3.1.1 release, so I've updated both mach-nix and pypi-deps-db to the latest version and verified that `torch==1.7.1` exists in the database:

```
$ cat ./wheel/37.json | jq '.torch.cp36."1.7.1"'
{
  "torch-1.7.1-cp36-cp36m-manylinux1_x86_64.whl": {
    "requires_dist": [
      "typing-extensions",
      "numpy",
      "dataclasses ; python_version < \"3.7\""
    ],
    "requires_python": ">=3.6.2"
  },
  "torch-1.7.1-cp36-cp36m-win_amd64.whl": "1.7.1@torch-1.7.1-cp36-cp36m-manylinux1_x86_64.whl",
  "torch-1.7.1-cp36-none-macosx_10_9_x86_64.whl": "1.7.1@torch-1.7.1-cp36-cp36m-manylinux1_x86_64.whl"
}
```

but still resolving fails:

```
$ nix-build ./repro.nix
...
Some requirements could not be resolved.
Top level requirements:
  torch==1.7.1
Providers:
  {'_default': 'wheel,sdist,nixpkgs',
   'gdal': 'nixpkgs',
   'graphql-core': 'sdist',
   'pip': 'nixpkgs,sdist',
   'setupmeta': 'wheel',
   'setuptools': 'nixpkgs',
   'wheel': 'nixpkgs,sdist'}
Mach-nix version: master
Python: 3.6.12
Cause: Requirements conflict: Requirement.parse('torch==1.7.1')
The requirements which caused the error:
  torch==1.7.1
```

If I switch back to `torch==1.7.0` it works. The deps info for that version looks like this:

```
$ cat ./wheel/37.json | jq '.torch.cp36."1.7.0"'
{
  "torch-1.7.0-cp36-cp36m-manylinux1_x86_64.whl": {
    "requires_dist": [
      "future",
      "typing-extensions",
      "dataclasses",
      "numpy"
    ],
    "requires_python": ">=3.6.1"
  },
  "torch-1.7.0-cp36-cp36m-win_amd64.whl": "1.7.0@torch-1.7.0-cp36-cp36m-manylinux1_x86_64.whl",
  "torch-1.7.0-cp36-none-macosx_10_9_x86_64.whl": "1.7.0@torch-1.7.0-cp36-cp36m-manylinux1_x86_64.whl"
}
```

I fail to see why 1.7.1 should fail when 1.7.0 works.

`./repro.nix` file:
```
let
  mach-nix = import (builtins.fetchGit {
    url = "https://github.com/DavHau/mach-nix/";
    ref = "refs/heads/master";
    rev = "afa94cb61ce68fad8688ce4f32fc7095962aa191";  # Thu Feb 18 15:56:16 2021 +0700
  }) {
    pypiDataRev = "2c5c3148a4a1c58e7e19dd76e57d331da4ce208d";  # Thu Feb 18 08:10:49 2021 +0000
    pypiDataSha256 = "0jfaxjl0clnj4f2q4w8xzqhmj1nvljw19sm3fb9cmfwns3s0b6kp";
    python = "python36";
  };

  customPython = mach-nix.mkPython {
    # torch==1.7.0 works, torch==1.7.1 fails.
    requirements = ''
      torch==1.7.1
    '';
  };
in
  customPython
```

**Comment by bjornfor** [comment]
So this was fixed in mach-nix-3.2.0. I'm worried though that the next time I need to upgrade pypiDataRev to get newer python packages, I'd hit the same issue :-/

**Comment by bjornfor** [comment]
I played a bit more with my `repro.nix` file from the OP and found that it works if I switch from "python36" to "python37". But I explicitly checked the pypi-deps-db for the "cp36" field (see OP), so I really don't understand why it wasn't found.

With the latest version (either using mach-nix-3.2.0 directly or updating `repro.nix` to use `pypiDataRev = "ab522f2d3255789f1ef97fa7c83d4342be156e67"` (same version as mach-nix-3.2.0 uses) then it works even works for "python36".

**Comment by bjornfor** [comment]
I had to dig some more.

Here are the two entries for torch==1.7.1 in failing vs working pypi-deps-db:

failing version (https://github.com/DavHau/pypi-deps-db/commit/2c5c3148a4a1c58e7e19dd76e57d331da4ce208d):

```
$ cat ./wheel/37.json | jq '.torch.cp36."1.7.1"'
{
  "torch-1.7.1-cp36-cp36m-manylinux1_x86_64.whl": {
    "requires_dist": [
      "typing-extensions",
      "numpy",
      "dataclasses ; python_version < \"3.7\""
    ],
    "requires_python": ">=3.6.2"
  },
  "torch-1.7.1-cp36-cp36m-win_amd64.whl": "1.7.1@torch-1.7.1-cp36-cp36m-manylinux1_x86_64.whl",
  "torch-1.7.1-cp36-none-macosx_10_9_x86_64.whl": "1.7.1@torch-1.7.1-cp36-cp36m-manylinux1_x86_64.whl"
}
```

working version (https://github.com/DavHau/pypi-deps-db/commit/ab522f2d3255789f1ef97fa7c83d4342be156e67):
```
$ cat ./wheel/37.json | jq '.torch.cp36."1.7.1"'
{
  "torch-1.7.1-cp36-cp36m-manylinux1_x86_64.whl": {
    "requires_dist": [
      "typing-extensions",
      "numpy",
      "dataclasses ; python_version < \"3.7\""
    ],
    "requires_python": ">=3.6.2"
  },
  "torch-1.7.1-cp36-cp36m-win_amd64.whl": "1.7.1@torch-1.7.1-cp36-cp36m-manylinux1_x86_64.whl",
  "torch-1.7.1-cp36-none-macosx_10_9_x86_64.whl": "1.7.1@torch-1.7.1-cp36-cp36m-manylinux1_x86_64.whl"
}
```

Those entries are identical, so the problem lies elsewhere in the dependency graph. (I don't know how to construct and/or diff that graph.)

I bisected the fix to be https://github.com/DavHau/pypi-deps-db/commit/ab522f2d3255789f1ef97fa7c83d4342be156e67, which happens to be the version that mach-nix pins in its flake.lock file for v3.2.0. Coincidence? I don't see anything particular with that commit, its a normal "wheel_update" like many before it. I'm confused.

**Comment by bjornfor** [comment]
torch 1.7.1 was released 2020-12-10. Why didn't it appear fully working in pypi-deps-db until 2021-03-10? (The top-level torch==1.7.1 entry was there at least on 2021-02-18, when I first encountered the issue, but then it only worked for python37, not python36.

**Comment by bjornfor** [comment]
I don't know how I came to the conclusion that this was fixed, I tried again now and it still fails.

Updated repro.nix file:

```nix
let
  mach-nix = import (builtins.fetchGit {
    url = "https://github.com/DavHau/mach-nix/";
    ref = "refs/tags/3.2.0";
  }) {
    python = "python36";
  };

  customPython = mach-nix.mkPython {
    # torch==1.7.0 works, torch==1.7.1 fails.
    # Switching from python36 to python37 makes torch=1.7.1 work though.
    requirements = ''
      torch==1.7.1
    '';
  };
in
  customPython
```

```
$ nix-build ./repro.nix
building '/nix/store/splnwa0drj0fxfpsfzhwwjhnmsi8xlif-mach_nix_file.drv'...

Some requirements could not be resolved.
Top level requirements:
  torch==1.7.1
Providers:
  {'_default': 'wheel,sdist,nixpkgs',
   'gdal': 'nixpkgs',
   'pip': 'nixpkgs,sdist',
   'setupmeta': 'wheel',
   'setuptools': 'nixpkgs',
   'wheel': 'nixpkgs,sdist'}
Mach-nix version: 3.2.0
Python: 3.6.13
Cause: Requirements conflict: Requirement.parse('torch==1.7.1')
The requirements which caused the error:
  torch==1.7.1

builder for '/nix/store/splnwa0drj0fxfpsfzhwwjhnmsi8xlif-mach_nix_file.drv' failed with exit code 1
error: build of '/nix/store/splnwa0drj0fxfpsfzhwwjhnmsi8xlif-mach_nix_file.drv' on 'ssh://nix-remote-build@bforsman.name' failed: builder for '/nix/store/splnwa0drj0fxfpsfzhwwjhnmsi8xlif-mach_nix_file.drv' failed with exit code 1
builder for '/nix/store/splnwa0drj0fxfpsfzhwwjhnmsi8xlif-mach_nix_file.drv' failed with exit code 1
error: build of '/nix/store/splnwa0drj0fxfpsfzhwwjhnmsi8xlif-mach_nix_file.drv' failed
(use '--show-trace' to show detailed location information)
```

**Comment by DavHau** [comment]
fixed via f139a8b68aed0646a42e1579f2db9ab4db43df31

**Comment by bjornfor** [comment]
Thanks!
